### ISSUE TITLE
v4 - no exported member QueryClient - with pnpm in a typescript monorepo

### ISSUE THREAD

**Issue by aestheticsdata** [TanStack/query#5267]
### Describe the bug

I've already opened an issue which has been closed : https://github.com/TanStack/query/issues/5203

But there is an issue. The issue is inside a monorepo with pnpm.

The QueryClient was available in v3. Switching to v4 I have the error : TS2724: '"@tanstack/react-query"' has no exported member named 'QueryClient'. Did you mean 'useQueryClient'?
But it's still there in the doc : https://tanstack.com/query/v4/docs/react/reference/QueryClient#queryclient

Trying to switch to useQueryClient
const queryClient = useQueryClient();
or const queryClient = useQueryClient({});
I have the error TypeError: Cannot read properties of null (reading 'useContext')
But reading the doc : https://tanstack.com/query/v4/docs/react/reference/useQueryClient , "Otherwise, defaultContext will be used."

### Your minimal, reproducible example

no sandbox

### Steps to reproduce

I created a monorepo with pnpm. I have two NextJS v13 apps inside : noumaios and themis.
install tanstack-query v4 inside noumaios with pnpm.
the tsconfig.js file has the option `preserveSymlinks` set to true (it builds without errors when `preserveSymlinks` is set to false).
try to build noumaios with `pnpm build`, and then the build can't complete and finish with the error '"@tanstack/react-query"' has no exported member named 'QueryClient'. Did you mean 'useQueryClient'?
See the screenshots below.

### Expected behavior

the build should pass without errors.

### How often does this bug happen?

None

### Screenshots or Videos

<img width="326" alt="Screenshot 2023-04-13 at 11 07 23" src="https://user-images.githubusercontent.com/453479/231744697-7cf3ba75-c98b-4819-a9c8-09568b46721f.png">

<img width="367" alt="Screenshot 2023-04-13 at 11 08 31" src="https://user-images.githubusercontent.com/453479/231744769-ef83c6c3-4281-4fb0-8756-8f9d4cdcfb6d.png">

<img width="979" alt="Screenshot 2023-04-13 at 11 09 09" src="https://user-images.githubusercontent.com/453479/231744796-b9f7f287-2ea6-4cab-bf4c-fa1d4d235f0c.png">

<img width="498" alt="Screenshot 2023-04-13 at 11 10 10" src="https://user-images.githubusercontent.com/453479/231744862-c9eb10e2-d8cd-43d8-9351-dda0005c134c.png">

<img width="1297" alt="Screenshot 2023-04-13 at 11 11 12" src="https://user-images.githubusercontent.com/453479/231744909-9bdd2919-471e-40e9-aa0f-4fa74781fed7.png">

<img width="819" alt="Screenshot 2023-04-13 at 11 15 26" src="https://user-images.githubusercontent.com/453479/231744960-b4a7f9e7-dd48-4682-9608-96676805e8c6.png">

<img width="843" alt="Screenshot 2023-04-13 at 11 15 18" src="https://user-images.githubusercontent.com/453479/231744986-cccd70cb-52ab-4523-a44b-441e0e7ce7ca.png">

### Platform

MacOS 11.6
Mac M1
NextJS v13

### Tanstack Query adapter

None

### TanStack Query version

v4.29.1

### TypeScript version

v5.0.4

**Comment by TkDodo** [comment]
opening a new issue with screenshots doesn't help though. What we need is a standalone reproduction. I said the same thing in the [issue you linked to](https://github.com/TanStack/query/issues/5203#issuecomment-1492881661):

> So you have to provide a reproduction please. Feel free to reopen if you can isolate the issue.

Now we have two issues with no reproduction 🤷 .

Just FYI, pnpm v8 has changed peerDependency resolution, which was one of the pain-points. Not sure which version you are using...

**Comment by aestheticsdata** [comment]
> opening a new issue with screenshots doesn't help though. What we need is a standalone reproduction. I said the same thing in the [issue you linked to](https://github.com/TanStack/query/issues/5203#issuecomment-1492881661):
> 
> > So you have to provide a reproduction please. Feel free to reopen if you can isolate the issue.
> 
> Now we have two issues with no reproduction 🤷 .
> 
> Just FYI, pnpm v8 has changed peerDependency resolution, which was one of the pain-points. Not sure which version you are using...

All the steps are detailed in order to be able to reproduce the problem. What can I do more to explain how to reproduce the issue ???
What the point of being able to include screenshot on a Github if you say that it doesn't help ???
I use pnpm v8, the latest version.

**Comment by aestheticsdata** [comment]
ok.
It's probably a peerDependency resolution conflict with typescript.
Thank you.

**Comment by aestheticsdata** [comment]
> so does that mean you tried to re-create an isolated reproduction and it worked?

As I explain in the "steps to reproduce" section in this issue, I created a test monorepo to isolate the issue as shown on the screenshots. And I explained that when `preserveSymLinks` in tsconfig.json is set to `true`, there is the issue, still with pnpm v8. As you told me that peerDependcy resolution with pnpm was a pain point, I assume that the issue has probably something to do with the typescript symbolic links resolution and the pnpm symbolic links resolution.
